### PR TITLE
Included a sanity check for systemctl edit

### PR DESCRIPTION
**Pull request by pengwyn**
Currently checks for an [Install] section in a drop-in replacment,
proper section names, and valid lvalue names in the variable
assignments. Warns about problems but doesn't take any action.

There's support for individual variable checks but currently nothing is
done.

Test by doing something like "systemctl edit console-getty.service" and pasting the following:

```
[Install]
Test=123
WantedBy=123

[Something]
blah=3

[Unit]
WantedBy=234
```

You should see:

```
[/etc/systemd/system/console-getty.service.d/override.conf:2] Unknown lvalue 'Test' in section 'Install'
[/etc/systemd/system/console-getty.service.d/override.conf:5] Unknown section 'Something'. Ignoring.
[/etc/systemd/system/console-getty.service.d/override.conf:9] Unknown lvalue 'WantedBy' in section 'Unit'
[/etc/systemd/system/console-getty.service.d/override.conf:0] 'Install' sections are not allowed in drop-in overrides. Make a full replacement using 'systemctl edit -l ...' instead.
```

### CODE REVIEW

**Comment by pengwyn** [comment]
Whoopes - this should be linked with issue #1832.

And apparently I don't know how to use Github properly... if I put resolves here will this work?
resolves #1832

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -6113,6 +6114,134 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+        assert(table == NULL);
+        warnings = (struct CheckSanityWarnings*)userdata;
+
+        /* There should be no install section in a drop-in override. */
+        if (!arg_full && strcmp(section,"Install") == 0 && !warnings->warn_about_install) {
+                warnings->warn_about_install = true;
+        }
+
+        /* Check the real lookup function first to see if the lvalue makes sense. */
+#if 0
```
Comment (zonque) [inline src/systemctl/systemctl.c]: No dead code, please.

**Comment by zonque** [comment]
I'm not sure whether we want that really. I see the benefit, but it also makes people believe we are doing a full check of the changes, which we don't and never will be able to do. But stopping at some arbitrary point in the process will be causing more confusion than it helps. Selling `systemctl edit` as simple wrapper around `$EDITOR` is a much more comprehensible deal IMO.

/cc @poettering

**Comment by pengwyn** [comment]
I see your point, although I would also argue that it's not clear to the user that `systemctl edit` is merely a wrapper around `$EDITOR`. I'd also argue that the current code doesn't make any statements saying "File x verified" - it just points out any obvious mistakes. But if you think that it might become bloated with checks for too many specific cases, maybe I can remove the `edit_check_parse` function and just stick to the generic tests that are done now?

The problem that motivated me for this was the way that the `[Install]` section was ignored in a drop-in override. It doesn't get mentioned in any of the logs so I had no idea what was going wrong.

I also removed that dead core you commented on and cleaned up some other parts.

**Comment by pengwyn** [comment]
By the way, this is only partially related, but I just realised that the return array from `strjoin` should be freed in the end. But doing a quick check through the repository, it seems like most uses of `strjoin` aren't freed. Is there some magic memory management going on that I'm not aware of?

Should all of these temporary strings have `_cleanup_free_` put in front of them?

**Comment by zonque** [comment]
> By the way, this is only partially related, but I just realised that the return array from strjoin should be freed in the end.

That's correct, yes. There are some other minor things in the patch, such as that curly braces (`{`, `}`) are not needed for single-statement blocks.

> But doing a quick check through the repository, it seems like most uses of strjoin aren't freed. Is there some magic memory management going on that I'm not aware of?
> 
> Should all of these temporary strings have _cleanup_free_ put in front of them?

Sure, unless they are returned to the caller, which transfers the responsibility to free the memory up the chain. If you spot any such leaks, please file a PR.

File: src/systemctl/systemctl.c [hunk 2]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+
+        struct CheckSanityWarnings * warnings = NULL;
+
+        assert(section);
+        assert(userdata);
+        assert(table == NULL);
+        warnings = (struct CheckSanityWarnings*)userdata;
+
+        /* There should be no install section in a drop-in override. */
+        if (!arg_full && streq(section,"Install") && !warnings->warn_about_install) {
```
Comment (zonque) [inline src/systemctl/systemctl.c:6154]: No curly braces for single-statement blocks.

File: src/systemctl/systemctl.c [hunk 3]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+         * Instead, use the nulstr output to manually parse the available
+         * options.
+         */
+        comparison = strjoin(section, ".", lvalue, NULL);
+        scan = load_fragment_gperf_nulstr;
+        while(*scan != '\0') {
+                if (streq(scan, comparison))
+                        break;
+                scan += strlen(scan)+1;
+        }
```
Comment (zonque) [inline src/systemctl/systemctl.c:6168]: As you're on it: add a newline here.

File: src/systemctl/systemctl.c [hunk 4]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+        const char * sections = 
+                "Unit\0"
+                "Service\0"
+                "Install\0";
+
+        struct CheckSanityWarnings warnings = { .warn_about_install = false};
+
+        assert(path);
+
+        file = fopen(path,"r");
```
Comment (zonque) [inline src/systemctl/systemctl.c:6199]: Stray newline here.

File: src/systemctl/systemctl.c [hunk 5]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+                         file,
+                         sections,
+                         edit_check_dummy_lookup,
+                         NULL,
+                         false, 
+                         false,
+                         true,
+                         &warnings);
+
+        if (warnings.warn_about_install) {
```
Comment (zonque) [inline src/systemctl/systemctl.c:6217]: See above.

File: src/systemctl/systemctl.c [hunk 6]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+ * for each line in a file.
+ */
+struct CheckSanityWarnings {
+        bool warn_about_install;
+};
+
+static int edit_check_parse(const char *unit, const char *filename, unsigned line, const char *section, unsigned section_line, const char *lvalue, int ltype, const char *rvalue, void *data, void *userdata) {
+
+        assert(unit);
+        assert(streq(unit,"CheckSanity"));
```
Comment (zonque) [inline src/systemctl/systemctl.c:6127]: Whitespace after comma, please.

File: src/systemctl/systemctl.c [hunk 7]
```diff
@@ -6113,6 +6114,123 @@ static int find_paths_to_edit(sd_bus *bus, char **names, char ***paths) {
...
+        /* Including the sections here manually. I tried to take them from
+         * service.c:service_vtable but that requires including libmount and
+         * woe ensued...
+         */
+        const char * sections = 
+                "Unit\0"
+                "Service\0"
+                "Install\0";
+
+        struct CheckSanityWarnings warnings = { .warn_about_install = false};
```
Comment (zonque) [inline src/systemctl/systemctl.c:6195]: Whitespace before `}`.

**Comment by zonque** [comment]
Before you address the issues I pointed out above, we should first conclude whether or not we want the entire thing.

**Comment by johannbg** [comment]
Since systemd does not provided a native built in editor itself then logic dictates that syntax checking/highlights and templates ( think new type unit files ) should be implement as an addon/extension in each $EDITOR either upstream with those editors or downstream with distribution that ship those editors ( people usually do that with VI(M) and probably EMACS as well )  not systemd itself.  

If and then when systemd provides it's own editor to manipulate those files then arguably it should provided those requested means. ( although I'm personally not foreseeing systemd ever providing it's own built in editor )

**Comment by filbranden** [comment]
I think a fair comparison is `visudo` which checks the contents of the sudoers file for basic syntax errors before replacing it. It will never prevent all mistakes but at least will help you catch typos... Do you think it helps to compare this feature to what `visudo` does?

**Comment by johannbg** [comment]
To me it's like comparing apple and oranges, visudo is an extension to sudo, an addon and deals with a single file on a single location with very limited number of options ( and it's checks are doing more behind the scene than just syntax checking ) . 

Systemd is a completely another beast in share number of configuration files and their location and entries and location of those entries in those files. 

If we speak just about type units then it's not enough for you to just syntax check them you also need to check if the line being added matches those that are applicable to just the type unit being edited and you also need to check if that line is being added to the correct section of the type unit file.

**Comment by pengwyn** [comment]
I'm still confused as to why these checks have to be complete. The `systemctl edit` tool is for convenience only, why can't it do some simple convenient checks at the same time?

After all, if `systemctl edit` were purely a wrapper around an `$EDITOR`, then wouldn't it be better to remove `systemctl edit` entirely and just put instructions in the docs? Something like "Include drop-in overrides by creating files in .../blahblah.d/"

**Comment by johannbg** [comment]
Because dealing with ignorance is never simple and half implementing anything wont do those people anything good. The tool would just end up being the limp leading the blind. 

If an individual is incapable of correctly spelling the entries to type unit files then he's equally prone to misspell the value's in those entries themselves ( which cannot be syntaxed checked and or verified ) more likely than not being incapable of adding the correct entries which are applicable to a spesific type unit file, correct entries based on his installed system ( SELinuxContext= vs AppArmorProfile= for example so systemctl edit would need to start verifying those entries in conjunction to syntax check them like visudo semi does )  and correctly placing those entries in the right [section] of the type units. 

The simple and less work for administrators of that nature would be if systemctl edit copied the entire existing unit file and presented that for administrators to edit visudo style instead of using drop-ins since doing so it would present the end user with already existing entries he could directly edit and or copy for that matter. 

The argument as practical as it might be for using drop-ins does not hold water in the real world.  ( drop-ins are great on paper grap on field kind of thing )

**Comment by pengwyn** [comment]
Fair enough. I still think it helps administrators to notify them of simple typos (they don't have to be ignorant, just hitting the wrong key) but I see your point.

The check for [Install] sections in drop-ins would be more useful because the behaviour is a more counter-intuitive. However, that would also be "fixed" by simply documenting it (which is something else I intend to add).

But you seem to be suggesting that drop-ins be removed from `systemctl edit` altogether. Should the `systemctl edit -l ...` behaviour be made the default? Or the only option?

**Comment by johannbg** [comment]
It's not possible to fully support fixing typos so you end up only supporting notifying/fixing only portion of the typo's right. 

For example execstart could be notified/fixed to become ExecStart but you cant fix types that come after it as in ExecStart=/mis/spelled/path/, ExecStart=/correct/spelled/path/foo $WRONGOPTION etc.

So you cannot solve this problem properly

systemctl edit -l should from my point of view be the default options and provides better end result then trying to implement syntax checking and drop-ins either be as an option or dropped altogether since an full copy of the type unit file provides end users already with complete working type unit ( and it should provided users with a template based one if he's writing a new one from scratch) thus that user can directly go to the relevant line he wants to change or add a new one ( which he most likely will do by copy/paste the relevant line from the relevant man page or from other documentation sources )  to the correct [section] of the unit file as opposed to having to write one from scratch as drop-in and know all the correct syntax, [sections] and what belongs where and spell them right.

**Comment by poettering** [comment]
I am not convinced this really should be done the way it is. Maybe instead do it more like a git commit hook thingy? I.e. invoke a set of executables from some dir, and only accept the thing when they all pass with 0. Then, install a symlink to systemd-analyze into that dir or so.

This way, people could even install their own hooks if they like (for example, one that adds extra checks, like requiring a Description= to be set or so).

I am really not convinced that the checks themselves should be built into systemctl, since they are necessarily incomplete, and people should not misunderstand that.

**Comment by johannbg** [comment]
@poettering I'm not following...  

If downstreams start implementing this then that will lead to further fragmentation between distributions and broken end user expectations and confusion ( unit works on distribution A that does not have an requirement but fails when he moves it to distribution B which has an requirement and before you know it administrator running something like puppet/chef will be knocking on the systemd door for explanation why that happens ) and upstreams have mixed feelings about supporting init system in general so I fail to see which target audience would be those that would use those hooks ( beside systemd itself ). 

Who do you foresee using this if it would get implemented?  

The best way and the least amount of work and maintenance to alleviate the current situation is stop defaulting the editor to create/edit drop-ins ( those will never work in practice anyway due to reasons externally to systemd ) and default it to copy/edit the original unit files instead.

**Comment by pengwyn** [comment]
I'm guessing from the discussion that this is not the right way to go with systemctl so I'm closing the request. Thanks for the comments on fixing up the code anyway @zonque.

Perhaps a new issue should be created, suggesting the change of default behaviour to be a full replacement? Or even to remove drop-ins completely?

**Comment by poettering** [comment]
Yeah, please open a new RFE for this. I think an implementation would be much simpler than this patch set, i.e. it might be enough to use the existing execute_directories() call on some new dir, after returning from the $EDITOR. The call would have to learn to care about non-zero exit codes of the executed binaries though, it currently ignores them fully.

**Pull request closed without merging**